### ISSUE TITLE
Should be able to use env var to set spring.cloud.stream.bindings.input

### ISSUE THREAD

**Issue by trisberg** [spring-cloud/spring-cloud-stream#105]
I'd like to be able to use:

    -e spring_cloud_stream_bindings_input=myqueue

when deploying to Docker/Lattice, this works for other Spring Boot config properties like

    -e spring_redis_host=myhost

**Comment by markfisher** [comment]
That should already work since it's exactly what the ReceptorModuleDeployer does:

https://github.com/spring-cloud/spring-cloud-data/blob/master/spring-cloud-data-module-deployers/spring-cloud-data-module-deployer-lattice/src/main/java/org/springframework/cloud/data/module/deployer/lattice/ReceptorModuleDeployer.java#L69

The parameters contain the binding values as populated here:
https://github.com/spring-cloud/spring-cloud-data/blob/master/spring-cloud-data-core/src/main/java/org/springframework/cloud/data/core/ModuleDefinitionBuilder.java#L87

The only difference is those are dot-delimited, but earlier they used underscores yet with uppercase (if that makes a difference).

**Comment by trisberg** [comment]
These work:

    java -Dspring.cloud.stream.bindings.output=myqueue -jar target/time-source-1.0.0.BUILD-SNAPSHOT-exec.jar

    java -jar target/time-source-1.0.0.BUILD-SNAPSHOT-exec.jar --
spring.cloud.stream.bindings.output=myqueue

but these don't:

    SPRING_CLOUD_STREAM_BINDINGS_OUTPUT=myqueue java -jar target/time-source-1.0.0.BUILD-SNAPSHOT-exec.jar

    spring_cloud_stream_bindings_output=myqueue java -jar target/time-source-1.0.0.BUILD-SNAPSHOT-exec.jar

**Comment by mbogoevici** [comment]
This is reproducible, but specific to `spring-cloud-stream-modules`, i.e. running the SCSM time-source:

   SPRING_CLOUD_STREAM_BINDINGS_OUTPUT=foo java -jar target/time-source-1.0.0.BUILD-SNAPSHOT-exec.jar 

will display the described behaviour, however, running the sample from within spring-cloud-stream:

   SPRING_CLOUD_STREAM_BINDINGS_OUTPUT=qux java -jar target/spring-cloud-stream-sample-source-1.0.0.BUILD-SNAPSHOT-exec.jar

will work as expected.

**Comment by mbogoevici** [comment]
Seems like the key difference is whether the value is set outside the OS environment. In other words, the SCS examples that still it set via the application.yml will allow overriding via environment, the others won't. Since all the other Boot overriding mechanisms (System properties, command-line arguments) work, I would say it's a Boot issue.

**Comment by mbogoevici** [comment]
Ok, it seems that this is caused by the way in which Spring bean OS environment binding happens for nested properties (`binding` in this case). The SystemEnvironmentPropertySource will be strictly queried for `SPRING_CLOUD_STREAM_BINDING` (which does not exist as such), and does not support nesting. A quick solution would be to replace the current defaulting mechanism that returns a value if the property is not set with a prepopulated PropertySource (which will be presumably overriden correctly).

**Comment by mminella** [comment]
For the record, I'm trying to get the equivalent to `http | log` running on cf and this doesn't work there either.  The following was tried (stream-cloud-streams-bus is the name of my redis service on CF):

````
$ cf push cc-http -p http-source/target/http-source-1.0.0.BUILD-SNAPSHOT-exec.jar --no-start
$ cf bs cc-http stream-cloud-streams-bus
$ cf se cc-http spring_cloud_stream_bindings_output http_out
$ cf start cc-http
$ cf push cc-log -p log-sink/target/log-sink-1.0.0.BUILD-SNAPSHOT-exec.jar --no-start
$ cf bs cc-log stream-cloud-streams-bus
$ cf se cc-log spring_cloud_stream_bindings_input http_out
$ cf start cc-log
```

**Comment by ilayaperumalg** [comment]
It turns out that for non-String property value mapping (list, map etc.,) spring-boot's `RelaxedDataBinder` would consider mapping the property name tokens into data-binding [**only** if the property names are delimited by `.`] (https://github.com/spring-projects/spring-boot/blob/master/spring-boot/src/main/java/org/springframework/boot/bind/RelaxedDataBinder.java#L502)

Hence, for this case, we need to specify the property this way:
```
SPRING_CLOUD_STREAM_BINDINGS.OUTPUT=myqueue java -jar target/time-source-1.0.0.BUILD-SNAPSHOT-exec.jar
spring_cloud_stream_bindings.output=myqueue java -jar target/time-source-1.0.0.BUILD-SNAPSHOT-exec.jar
```
Note the `.` for `bindings.output` in both the cases for the property name after prefix `spring.cloud.stream` on `ChannelBindingProperties`.

Even if we set the system property this way:
```
java -Dspring_cloud_stream_bindings_output=myqueue -jar target/time-source-1.0.0.BUILD-SNAPSHOT-exec.jar
```
that wouldn't work because of `_` for `bindings_output`.

I believe it makes sense to have such a differentiation (to have `.` delimited property names for collections etc.,) as the property name itself could have `_` for direct string mapping.

**Comment by markfisher** [comment]
but that means it's unusable as a standard env var:

```
$ export FOO_BAR.oops
-bash: export: `FOO_BAR.oops': not a valid identifier
```

**Comment by trisberg** [comment]
Fixed by @mbogoevici with commit 83123c3b97ee731e448d99df26b0f0bf8e517c81
